test/functional/test_framework/test_node.py

```diff
@@ -267,6 +284,96 @@ def assert_debug_log(self, expected_msgs):
+        tool like `hotspot`<https://github.com/KDAB/hotspot>.
+
+        To see useful textual output, run
+
+            perf report -i /path/to/datadir/node-0.perf.data --stdio | c++filt | less
+
+        See also:
+
+            Installing perf: https://askubuntu.com/q/50145
+            Perf examples: http://www.brendangregg.com/perf.html
+
+        Args:
+
+            profile_name (str): This string will be appended to the
+                profile data filename generated by perf.
+
+        Usage:
+
+            with node.profile_with_perf("send-big-msgs"):
+                # Perform activity on the node you're interested in profiling, e.g.:
+                for _ in range(10000):
+                    node.p2p.send_message(some_large_message)
+        """
+        subp = self._start_perf(profile_name)
+
+        yield
+
+        if subp:
+            self._stop_perf(profile_name)
+
+    def _start_perf(self, profile_name):
+        """Start a perf process to profile this node.
+
+        Returns the subprocess running perf.
+        """
+        subp = None
+
+        def test_success(cmd):
+            try:
+                return subprocess.run(
+                    cmd, shell=True,
+                    stderr=subprocess.DEVNULL, stdout=subprocess.DEVNULL).returncode == 0
+            except (FileNotFoundError, subprocess.CalledProcessError):
+                self.log.warning("Unable to execute command '{}'".format(cmd))
+                return False
+
+        if not test_success('readelf -S {} | grep .debug_str'.format(shlex.quote(self.binary))):
+            self.log.warning(
+                "perf output won't be very useful without debug symbols compiled into bitcoind")
+
+        if not test_success('which perf'):
+            self.log.warning("Can't profile with perf; must install perf-tools")
+        else:
+            output_path = self._get_perf_data_path(profile_name)
+
+            cmd = [
+                'perf', 'record',
+                '-g',                     # Record the callgraph.
+                '--call-graph', 'dwarf',  # Compatibility for gcc's --fomit-frame-pointer.
+                '-F', '101',              # Sampling frequency in Hz.
+                '-p', str(self.process.pid),
+                '-o', output_path,
+            ]
+            subp = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            self.perf_subprocesses[profile_name] = subp
+
+        return subp
+
+    def _stop_perf(self, profile_name):
+        """Stop (and pop) a perf subprocess."""
+        subp = self.perf_subprocesses.pop(profile_name)
+        subp.terminate()
+        subp.wait()
+        self.log.info("Wrote perf output to '{}'".format(
+            self._get_perf_data_path(profile_name)))
+
+    def _get_perf_data_path(self, profile_name):
+        """Return a suitable path for perf data given some descriptive name."""
+        suffix = "-{}".format(profile_name) if profile_name else ""
```

Comment:
profile_name is always given, so no need for the else?

Reply:
It's an empty string when we specify `--perf` to the runner: https://github.com/bitcoin/bitcoin/blob/358e757d566d5f659901acc6f15fa9eb6eb3e741/test/functional/test_framework/test_node.py#L184

Reply:
Wouldn't that overwrite the perf file on restart?

Reply:
Ah, good catch. Will verify & fix.